### PR TITLE
feat: add directory_path rule and DirectoryPathInfo for specifying files within tree artifacts

### PR DESCRIPTION
**Pull request by gregmagolan**
Used by copy_to_directory and copy_file

### CODE REVIEW

**Comment by gregmagolan** [comment]
Still need to add a test that for copy_file, but the implementation and test for copy_to_directory are ready for reveiw

File: lib/private/directory_path.bzl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+def _directory_path(ctx):
+    if not ctx.file.directory.is_directory:
+        fail("directory attribute must be created with Bazel declare_directory (TreeArtifact)")
+    return [DirectoryPathInfo(path = ctx.attr.path, directory = ctx.file.directory)]
+
+directory_path = rule(
+    doc = """Provide DirectoryPathInfo to reference some path within a directory.
+
+        Otherwise there is no way to give a Bazel label for it.""",
```
Comment (kormide) [inline lib/private/directory_path.bzl:23]: The .md preview seems to interpret this indent as a code block.

File: lib/private/copy_to_directory.bzl [hunk 2]
```diff
@@ -112,43 +35,54 @@ def _longest_match(subject, tests, allow_partial = False):
                 high_score = score
     return match
 
-def _output_path(ctx, src):
+# src can either be a File a target with a DirectoryPathInfo
```
Comment (kormide) [inline lib/private/copy_to_directory.bzl:38]: nit: "or a target"

File: lib/tests/copy_to_directory/BUILD.bazel [hunk 3]
```diff
@@ -175,6 +176,50 @@ copy_to_directory(
     ],
 )
 
+# Case 9: directory_path srcs & replace_prefixes (with external repo)
+copy_to_directory(
```
Comment (kormide) [inline lib/tests/copy_to_directory/BUILD.bazel:180]: Do any sources in this example use directory paths? I don't see any `directory_path` declarations.

**Comment by kormide** [comment]
Very cool :+1:

File: lib/tests/copy_to_directory/BUILD.bazel (hunk 3 above)
Comment (gregmagolan) [reply lib/tests/copy_to_directory/BUILD.bazel:180]: Just the implicit version of it in this case 9

```
{":a": "a"},
```

for example is equivalent to

```
directory_path(
   name = ...
   directory = ":a",
   path = "a",
)
```

File: lib/private/copy_to_directory.bzl (hunk 2 above)
Comment (gregmagolan) [reply lib/private/copy_to_directory.bzl:38]: 🙏

File: lib/tests/copy_to_directory/BUILD.bazel (hunk 3 above)
Comment (gregmagolan) [reply lib/tests/copy_to_directory/BUILD.bazel:180]: it gets expanded in the copy_to_directory macro

File: lib/private/directory_path.bzl (hunk 1 above)
Comment (gregmagolan) [reply lib/private/directory_path.bzl]: Thanks for catching

**Comment by gregmagolan** [comment]
copy_file fixed up and test added

**Review by kormide — APPROVED** [review]
lgtm

File: lib/private/directory_path.bzl [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            allow_single_file = True,
+        ),
+        "path": attr.string(
+            doc = "path relative to the directory",
+            mandatory = True,
+        ),
+    },
+)
+
+def maybe_directory_path(name, maybe_directory_path):
```
Comment (gregmagolan) [inline lib/private/directory_path.bzl:38]: to_directory_path

File: lib/copy_to_directory.bzl [hunk 5]
```diff
@@ -1,22 +1,141 @@
...
+
+            directory_path() targets are accepted for specifying files/folders to copy
+            from within a directory artifact (TreeArtifact). These may also be specified
+            with the convienence syntax of a dict with a single directory artifact key and
+            path within the directory. For example,
+
+            ```
+            copy_to_directory(
+                name = "dir",
+                srcs = [
```
Comment (gregmagolan) [inline lib/copy_to_directory.bzl:41]:
```
srcs = [
    "//path/to/some/file:target",
] + dir_paths({
   "//path/to/some/directory:target": "path/within/the/directory/artifact1",
   "//path/to/some/directory:target": "path/within/the/directory/artifact2",
})
```

**Pull request merged**